File: src/world/LightEngine.ts
import { CHUNK_HEIGHT, CHUNK_SIZE, lightEmission, lightPasses } from './blocks';
import type { Chunk } from './Chunk';

const SKY_MAX = 15;

export type LightWorld = {
  getBlock(wx: number, y: number, wz: number): number;
  getChunk(cx: number, cz: number): Chunk | undefined;
};

/**
 * Fast column lighting (no BFS). Sky = full daylight in open columns;
 * block emitters get a tiny local falloff. Orders of magnitude cheaper than flood-fill.
 */
export function rebuildChunkLights(chunk: Chunk, _world: LightWorld): void {
  chunk.skyLight.fill(0);
  chunk.blockLight.fill(0);

  const emitters: number[] = [];

  for (let z = 0; z < CHUNK_SIZE; z++) {
    for (let x = 0; x < CHUNK_SIZE; x++) {
      let sky = SKY_MAX;
      for (let y = CHUNK_HEIGHT - 1; y >= 0; y--) {
        const i = chunk.index(x, y, z);
        const b = chunk.voxels[i]!;
        if (!lightPasses(b)) {
          sky = 0;
          chunk.skyLight[i] = 0;
        } else {
          // Soft cave fill so sealed air isn't pitch black (still well below daylight).
          chunk.skyLight[i] = sky > 0 ? sky : 5;
        }
        const em = lightEmission(b);
        if (em > 0) {
          chunk.blockLight[i] = em;
          emitters.push(x, y, z, em);
        }
      }
    }
  }

    // Local torch/crystal glow (radius ≤ 8) — no cross-chunk BFS.
  for (let e = 0; e < emitters.length; e += 4) {
    const ex = emitters[e]!;
    const ey = emitters[e + 1]!;
    const ez = emitters[e + 2]!;
    const em = emitters[e + 3]!;
    const r = Math.min(8, em);
    for (let dy = -r; dy <= r; dy++) {
      const y = ey + dy;
      if (y < 0 || y >= CHUNK_HEIGHT) continue;
      for (let dz = -r; dz <= r; dz++) {
        const z = ez + dz;
        if (z < 0 || z >= CHUNK_SIZE) continue;
        for (let dx = -r; dx <= r; dx++) {
          const x = ex + dx;
          if (x < 0 || x >= CHUNK_SIZE) continue;
          const dist = Math.abs(dx) + Math.abs(dy) + Math.abs(dz);
          if (dist === 0 || dist > r) continue;
          const i = chunk.index(x, y, z);
          if (!lightPasses(chunk.voxels[i]!)) continue;
          const lvl = em - dist;
          if (lvl > chunk.blockLight[i]!) chunk.blockLight[i] = lvl;
        }
      }
    }
  }

  chunk.lightsDirty = false;
}

export function sampleLight(world: LightWorld, wx: number, y: number, wz: number): number {
  if (y < 0) return 0;
  if (y >= CHUNK_HEIGHT) return SKY_MAX;
  const cx = Math.floor(wx / CHUNK_SIZE);
  const cz = Math.floor(wz / CHUNK_SIZE);
  const chunk = world.getChunk(cx, cz);
  if (!chunk) return 8;
  const lx = wx - cx * CHUNK_SIZE;
  const lz = wz - cz * CHUNK_SIZE;
  const i = chunk.index(lx, y, lz);
  return Math.max(chunk.skyLight[i]!, chunk.blockLight[i]!);
}

File: Assets/VYTHERA/Voxel/Lighting/LightEngine.cs
using System;
using System.Collections.Generic;
using VYTHERA.Voxel.Data;
using VYTHERA.Voxel.Storage;

namespace VYTHERA.Voxel.Lighting
{
    public static class LightEngine
    {
        public const byte SkyMax = 15;

        private struct LightNode
        {
            public int X;
            public int Y;
            public int Z;
            public byte Level;
        }

        public static void RebuildChunkLights(ChunkData chunk)
        {
            Array.Clear(chunk.SkyLight, 0, chunk.SkyLight.Length);
            Array.Clear(chunk.BlockLight, 0, chunk.BlockLight.Length);

            var emitters = new List<LightNode>(64);

            for (int z = 0; z < VoxelConstants.ChunkSize; z++)
            {
                for (int x = 0; x < VoxelConstants.ChunkSize; x++)
                {
                    byte sky = SkyMax;
                    for (int y = VoxelConstants.ChunkHeight - 1; y >= 0; y--)
                    {
                        int i = ChunkData.GetIndex(x, y, z);
                        byte b = chunk.Voxels[i];
                        if (!BlockUtility.LightPasses(b))
                        {
                            sky = 0;
                            chunk.SkyLight[i] = 0;
                        }
                        else
                        {
                            chunk.SkyLight[i] = sky > 0 ? sky : (byte)5;
                        }

                        byte em = BlockUtility.LightEmission(b);
                        if (em > 0)
                        {
                            chunk.BlockLight[i] = em;
                            emitters.Add(new LightNode { X = x, Y = y, Z = z, Level = em });
                        }
                    }
                }
            }

            // Local torch/crystal/lava glow (radius <= 8)
            int emitterCount = emitters.Count;
            for (int e = 0; e < emitterCount; e++)
            {
                var node = emitters[e];
                int r = Math.Min(8, (int)node.Level);

                for (int dy = -r; dy <= r; dy++)
                {
                    int y = node.Y + dy;
                    if ((uint)y >= VoxelConstants.ChunkHeight) continue;

                    for (int dz = -r; dz <= r; dz++)
                    {
                        int z = node.Z + dz;
                        if ((uint)z >= VoxelConstants.ChunkSize) continue;

                        for (int dx = -r; dx <= r; dx++)
                        {
                            int x = node.X + dx;
                            if ((uint)x >= VoxelConstants.ChunkSize) continue;

                            int dist = Math.Abs(dx) + Math.Abs(dy) + Math.Abs(dz);
                            if (dist == 0 || dist > r) continue;

                            int i = ChunkData.GetIndex(x, y, z);
                            if (!BlockUtility.LightPasses(chunk.Voxels[i])) continue;

                            byte lvl = (byte)(node.Level - dist);
                            if (lvl > chunk.BlockLight[i])
                            {
                                chunk.BlockLight[i] = lvl;
                            }
                        }
                    }
                }
            }

            chunk.IsLightsDirty = false;
        }
    }
}
